massa-bootstrap/src/server_binder.rs

```diff
@@ -117,48 +133,48 @@ impl<D: Duplex + 'static> BootstrapServerBinder<D> {
     /// 5. runs the passed in closure (typically a custom logging msg)
     ///
     /// consumes the binding in the process
-    pub(crate) fn close_and_send_error<F>(
-        mut self,
-        server_outer_rt_hnd: Handle,
-        msg: String,
-        addr: SocketAddr,
-        close_fn: F,
-    ) where
+    pub(crate) fn close_and_send_error<F>(mut self, msg: String, addr: SocketAddr, close_fn: F)
+    where
         F: FnOnce() + Send + 'static,
     {
         thread::Builder::new()
             .name("bootstrap-error-send".to_string())
             .spawn(move || {
                 let msg_cloned = msg.clone();
-                let err_send =
-                    server_outer_rt_hnd.block_on(async move { self.send_error(msg_cloned).await });
+                let err_send = self.send_error_timeout(msg_cloned);
                 match err_send {
-                    Err(_) => error!(
+                    Err(BootstrapError::IoError(e)) if e.kind() == ErrorKind::TimedOut => error!(
                         "bootstrap server timed out sending error '{}' to addr {}",
                         msg, addr
                     ),
-                    Ok(Err(e)) => error!("{}", e),
-                    Ok(Ok(_)) => {}
+                    Err(e) => error!("{}", e),
+                    Ok(_) => {}
                 }
                 close_fn();
             })
             // the non-builder spawn doesn't return a Result, and documentation states that
             // it's an error at the OS level.
             .unwrap();
     }
-    pub async fn send_error(
-        &mut self,
-        error: String,
-    ) -> Result<Result<(), BootstrapError>, Elapsed> {
-        tokio::time::timeout(
-            self.write_error_timeout.into(),
-            self.send(BootstrapServerMessage::BootstrapError { error }),
+    pub fn send_error_timeout(&mut self, error: String) -> Result<(), BootstrapError> {
+        self.send_timeout(
+            BootstrapServerMessage::BootstrapError { error },
+            Some(self.write_error_timeout.to_duration()),
         )
-        .await
+        .map_err(|e| match e {
+            BootstrapError::IoError(e) if e.kind() == ErrorKind::WouldBlock => {
+                BootstrapError::TimedOut(ErrorKind::TimedOut.into())
+            }
+            e => e,
+        })
     }
 
     /// Writes the next message. NOT cancel-safe
-    pub async fn send(&mut self, msg: BootstrapServerMessage) -> Result<(), BootstrapError> {
+    pub fn send_timeout(
```

Comment:
Proper serializer here would be an improvement along with the `next_timeout` deser, as a follow-up

Reply:
#3842